### ISSUE TITLE
State history no longer keeps a local history 

### ISSUE THREAD

**Issue by ceesem** [seung-lab/neuroglancer#480]
It is useful to have some history of your state at reasonably frequent intervals in case you accidentally deselect your neuron of interest (which should not be as difficult to retrieve than it is...) or you accidentally move off of the neuroglancer website. After the introduction of the local storage/local_id state storage, such save states were part of the history under the $ (Save History) button. However, at some point this feature either stopped working or was intentionally removed and now the Save History button only stores manual saves.

**Comment by ogewan** [comment]
In order to resolve some issues, Save state was change to reuse the ID per tab. The ID only changes using a URL that doesn't have a local ID or if two tabs are using the same local ID. As a result, on state changes it gets updates the same entry rather than push a new one.

For save history, returning that old functionality shouldn't be a problem. The main issue is that save history is now just storing the URL to the manually saved states. Going back means at worst storing the whole state for every change and local memory will quickly be filled (even faster if the state has many annotations). Maybe some combination of storing a diff and using a timer to store state will work.

Once annotations are moved out of the state, it won't be as serious but that's a long term project.

**Comment by ceesem** [comment]
> Going back means at worst storing the whole state for every change and local memory will quickly be filled (even faster if the state has many annotations). Maybe some combination of storing a diff and using a timer to store state will work.

This feels like a good solution to me. The current solution with a static local id per tab makes total sense, and adding a side save in the background every 5 minutes or so (configurable in the settings) would be great. My colleagues have gotten good at using the server-based save feature for states that they want to keep, so this extra local save is basically as a revert/undo if something goes awry. As such, getting rid of the oldest files when storage runs out would be fine. I did a quick check and a state with 1000 annotations points is ~90KB once its compressed (which could be done with [something like this](http://undefinedmethod.com/blog/storing-compressed-json-objects-in-local-storage-with-lz-string/)), so there should be room for several history files even in of bulky states.

I'm a bit confused by what you mean with "once annotations are moved out of the state..." — I can see annotations also coming from a database in addition to user input and (sometimes quite large) programmatically generated neuroglancer states, but that will be an additional feature not one one that replaces other methods. Is there another storage angle that is in the works?

**Comment by ogewan** [comment]
Iirc Dynamic Annotations (read/write to server) would replace locally stored annotations.

**Comment by ceesem** [comment]
Yeah, Dynamic Annotations were what I was referring to above about the database. They will be extremely useful, but not the only game in town.
